includes/widgets/class-wc-widget-layered-nav.php

```diff
@@ -425,19 +442,53 @@ protected function get_filtered_term_product_counts( $term_ids, $taxonomy, $quer
 		}
 
 		if ( ! isset( $cached_counts[ $query_hash ] ) ) {
-			// phpcs:disable WordPress.DB.PreparedSQL.NotPrepared
-			$results                      = $wpdb->get_results( $query_sql, ARRAY_A );
-			$counts                       = array_map( 'absint', wp_list_pluck( $results, 'term_count', 'term_count_id' ) );
+			$counts                       = $this->get_term_product_counts_from_queries( $main_query_sql, $query_sql_for_attributes_with_any_value );
 			$cached_counts[ $query_hash ] = $counts;
 			if ( true === $cache ) {
 				set_transient( 'wc_layered_nav_counts_' . sanitize_title( $taxonomy ), $cached_counts, DAY_IN_SECONDS );
 			}
-			// phpcs:enable WordPress.DB.PreparedSQL.NotPrepared
 		}
 
 		return array_map( 'absint', (array) $cached_counts[ $query_hash ] );
 	}
 
+	/**
+	 * Get the count of terms for products, using a set of SQL queries that are return pairs of product id - term id.
+	 *
+	 * @param string ...$queries SQL queries to use, each must return a "product_id" column and a "terms_count_id" column.
+	 *
+	 * @return array An array where the keys are term ids, and the values are term counts.
+	 */
+	private function get_term_product_counts_from_queries( ...$queries ) {
+		global $wpdb;
+
+		$total_counts = null;
+
+		foreach ( $queries as $query ) {
+			$query = "
+				SELECT COUNT(DISTINCT(product_id)) AS term_count, term_count_id FROM (
+					{$query}
+				) AS x GROUP BY term_count_id";
+
+			$results = $wpdb->get_results( $query, ARRAY_A ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
```

Comment:
lets cache here instead of in callee function, so that both the queries can be cached separately

Reply:
Do we really need to cache the queries separately? We have split the original query in two for performance reasons, but conceptually they are still a single one; they will always be used together and with the same set of parameters.